### PR TITLE
Enforce well formedness for type alias impl trait's hidden type

### LINKED ISSUE #84657: Type alias impl trait doesn’t properly enforce outlives relations.
```rust
#![feature(min_type_alias_impl_trait)]

use std::marker::PhantomData;

trait ProofForConversion<'a, 'b> {
    fn convert<T: ?Sized>(_: PhantomData<Self>, r: &'a T) -> &'b T;
}

impl<'a, 'b> ProofForConversion<'a, 'b> for &'b &'a () {
    fn convert<T: ?Sized>(_: PhantomData<Self>, r: &'a T) -> &'b T {
        r
    }
}

type Converter<'a, 'b> = impl ProofForConversion<'a, 'b>;

// Even _defining_use with an explicit `'a: 'b` compiles fine, too.
fn _defining_use<'a, 'b>(x: &'b &'a ()) -> Converter<'a, 'b> {
    x
}

fn extend_lifetime<'a, 'b, T: ?Sized>(x: &'a T) -> &'b T {
    Converter::<'a, 'b>::convert(PhantomData, x)
}

fn main() {
    let d;
    {
        let x = "Hello World".to_string();
        d = extend_lifetime(&x);
    }
    println!("{}", d);
}
```
```
�C
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0cc5cc0200664652a55a3387404bc5a1))

I would’ve expected rustc to complain about the `_defining_use` not matching the `type Converter<'a, 'b>` with a suggestion to turn it into `type Converter<'a: 'b, 'a>`.

@rustbot modify labels: A-typesystem, A-lifetimes, A-traits, A-impl-trait, F-type_alias_impl_trait, T-compiler, requires-nightly
and someone please add “I-unsound 💥” (and/or tell me if there is a way to do it myself).

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #84657

This was not an issue with return-position-impl-trait because the generic bounds of the function are the same as those of the opaque type, and the hidden type must already be well formed within the function.

With type-alias-impl-trait the hidden type could be defined in a function that has *more* lifetime bounds than the type alias. This is fine, but the hidden type must still be well formed without those additional bounds.

### CODE REVIEW

File: compiler/rustc_data_structures/src/vec_map.rs [hunk 1]
```diff
@@ -144,7 +144,7 @@ impl<'a, K, V> IntoIterator for &'a VecMap<K, V> {
     }
 }
 
-impl<'a, K, V> IntoIterator for &'a mut VecMap<K, V> {
+impl<'a, K: 'a, V: 'a> IntoIterator for &'a mut VecMap<K, V> {
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_data_structures/src/vec_map.rs:147]: Should we be concerned for breakage and get a lang team signoff?

**Comment by cjgillot** [comment]
I am probably not the right reviewer for type-related stuff.
r? @compiler-errors

File: compiler/rustc_data_structures/src/vec_map.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_data_structures/src/vec_map.rs:147]: Nope, this is just a nightly feature. We can add more implicit magic later

[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_data_structures/src/vec_map.rs:147]: yeah, i think this is something we can fix later

**Review by compiler-errors — APPROVED** [review]
loooks good, and it's behind a feature-gate anyways so i'm not worried about the regression in associated TAIT.

@bors r+

**Comment by compiler-errors** [comment]
oh lol bors does not respond to approval messages it seems

@bors r+
@bors rollup=never (so we can bisect this better later, feel free to change rollup status as you see fit)

**Comment by compiler-errors** [comment]
Same failure as referenced in https://github.com/rust-lang/rust/pull/95775#issuecomment-1092400718 ?

@bors retry

**Pull request merged**